Detect and use __builtin_assume for ASSUME

khwilliamson opened the pull request:
Commit 5d5b9c460e2a06563d2b5e35a1a79991460696eb fixed the definition of
ASSUME for some purposes, but broke it on Windows.  This commit should
fix that.  It also adds support for clang's __builtin_assume()

richardleach commented:
This still leaves `HAS_BUILTIN_UNREACHABLE_` undefined on non-DEBUGGING builds and doesn't consider _clang_'s `__builtin_assume`. Supporting both in non-DEBUGGING builds seems desirable for providing hints to compiler optimizers.

richardleach commented:
As per IRC, pls ignore my comment!

richardleach commented:
Thumbs up from me, but this seems to emit a new (harmless) warning on _clang_ builds:

```
clang-7 -c -DPERL_CORE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement gv.c
gv.c:289:12: warning: the argument to '__builtin_assume' has side effects that will be discarded [-Wassume]
    ASSUME(CvNAME_HEK(cv));
           ^~~~~~~~~~~~~~
./perl.h:3890:39: note: expanded from macro 'ASSUME'
#  define ASSUME(x)  __builtin_assume(x)
                                      ^
1 warning generated.
```

hvds commented:
> Thumbs up from me, but this seems to emit a new (harmless) warning on _clang_ builds:
> 
> ```
> clang-7 -c -DPERL_CORE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement gv.c
> gv.c:289:12: warning: the argument to '__builtin_assume' has side effects that will be discarded [-Wassume]
>     ASSUME(CvNAME_HEK(cv));

That's odd, the only thing I can imagine that coming from is the two MUTABLE_PTR calls in the expansion:
```
#if defined(PERL_USE_GCC_BRACE_GROUPS)
#  define MUTABLE_PTR(p) ({ void *p_ = (p); p_; })
#else
#  define MUTABLE_PTR(p) ((void *) (p))
#endif
```

It should be possible to check if that's so by building with `-DPERL_GCC_BRACE_GROUPS_FORBIDDEN` and seeing if that makes the warning go away.

richardleach commented:
I built with:
`./Configure -des -Dusedevel -Dcc=clang-7 -DPERL_GCC_BRACE_GROUPS_FORBIDDEN && make perl`
Still throws the warning:
```
clang-7 -c -DPERL_CORE -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2 -std=c89 -O2 -Wall -Werror=pointer-arith -Wextra -Wc++-compat -Wwrite-strings -Werror=declaration-after-statement gv.c
gv.c:289:12: warning: the argument to '__builtin_assume' has side effects that will be discarded [-Wassume]
    ASSUME(CvNAME_HEK(cv));
           ^~~~~~~~~~~~~~
./perl.h:3890:39: note: expanded from macro 'ASSUME'
#  define ASSUME(x)  __builtin_assume(x)
                                      ^
1 warning generated.
```

richardleach commented:
Is it because `CvNAME_HEK` is an inlined function, rather than just a statement? This looks related: https://www.mail-archive.com/bug-gnulib@gnu.org/msg39071.html

I've got a later version of clang in a different VM, will boot it up.

richardleach commented:
The warning goes away - for clang-11 - with either this change (which I presume isn't portable as-is):
```
# git diff cv.h
diff --git a/cv.h b/cv.h
index 435dee6261..32fa2b23bd 100644
--- a/cv.h
+++ b/cv.h
@@ -222,7 +222,7 @@ See L<perlguts/Autoloading with XSUBs>.
 /* Flags for newXS_flags  */
 #define XS_DYNAMIC_FILENAME    0x01    /* The filename isn't static  */
 
-PERL_STATIC_INLINE HEK *
+PERL_STATIC_INLINE HEK __attribute__ ((__const__)) *
 CvNAME_HEK(CV *sv)
 {
     return CvNAMED(sv)
```
or this sort of change:
```
# git diff gv.c
diff --git a/gv.c b/gv.c
index 10b729bf02..a169cd6aea 100644
--- a/gv.c
+++ b/gv.c
@@ -283,10 +283,11 @@ Perl_cvgv_from_hek(pTHX_ CV *cv)
 {
     GV *gv;
     SV **svp;
+    HEK* rltest = CvNAME_HEK(cv);
     PERL_ARGS_ASSERT_CVGV_FROM_HEK;
     assert(SvTYPE(cv) == SVt_PVCV);
     if (!CvSTASH(cv)) return NULL;
-    ASSUME(CvNAME_HEK(cv));
+    ASSUME(rltest);
     svp = hv_fetchhek(CvSTASH(cv), CvNAME_HEK(cv), 0);
     gv = MUTABLE_GV(svp && *svp ? *svp : newSV(0));
     if (!isGV(gv))
```

hvds commented:
> Is it because `CvNAME_HEK` is an inlined function, rather than just a statement? This looks related: https://www.mail-archive.com/bug-gnulib@gnu.org/msg39071.html

Ah yes, in light of that message that seems quite likely.

> The warning goes away - for clang-11 - with either this change (which I presume isn't portable as-is):
[snip]

I think the first priority should be to concentrate on making `ASSUME()` safe, if that's possible; it does not seem wise to me to use it if we're also going to have to change every inline function that might get invoked by some use of it.

hvds commented:
> I think the first priority should be to concentrate on making `ASSUME()` safe, if that's possible

Possibly using the approach within that message:
```
+/* Use a temporary variable, to avoid a clang warning
+   "the argument to '__builtin_assume' has side effects that will be discarded"
+   if R contains invocations of functions not marked as 'const'.  */
+# define assume(R) \
+    ((void) ({ __typeof__ (R) temp = (R); \
+               __builtin_assume (temp); }))
```

khwilliamson commented:
I think that this PR should be split into two, with the addition of the clang portion separated.  Then we get ASSUME anyway on compilers that have an UNREACHABLE method.  That portion seems uncontroversial

khwilliamson committed:
Fix ASSUME on clang

Avoid a spurious clang warning by turning off that particular warning
around this operation.

The idea given in
https://www.mail-archive.com/bug-gnulib@gnu.org/msg39071.html
doesn't work on bitfields

This fixes GH #18839

khwilliamson commented:
bf578026aedea7702aef5f819d8a56da9d959fb2 implements the non-problematic portions of this

jkeenan commented:
> [bf57802](https://github.com/Perl/perl5/commit/bf578026aedea7702aef5f819d8a56da9d959fb2) implements the non-problematic portions of this

@khwilliamson, what (presumably problematic) parts remain in this pull request.  Would they be better off in a fresh p.r.?

khwilliamson commented:
@richardleach Please check if this now works for you.  I followed @hvds advice

Thanks for @jkeenan for pointing out this languishing ticket

jkeenan commented:
> @richardleach Please check if this now works for you. I followed @hvds advice
> 

When I try building this branch on FreeBSD-12 (my customary threaded build, `clang10`), `make` very quickly fails:
```
[perlmonger: perl] $ gitcurr
pr3
[perlmonger: perl] $ git describe
v5.37.1-83-g3d480457fa
[perlmonger: perl] $ make
echo @`sh  cflags "optimize='-O2 -pipe -fstack-protector -fno-strict-aliasing'" opmini.o`  -DPERL_IS_MINIPERL -DPERL_EXTERNAL_GLOB opmini.c
@cc -c -DPERL_CORE -DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -std=c99 -O2 -pipe -fstack-protector -fno-strict-aliasing -Wall -Werror=pointer-arith -Werror=vla -Wextra -Wno-long-long -Wno-declaration-after-statement -Wc++-compat -Wwrite-strings -DPERL_IS_MINIPERL -DPERL_EXTERNAL_GLOB opmini.c
In file included from op.c:163:
In file included from ./perl.h:7494:
./sv_inline.h:456:10: error: invalid application of 'typeof' to bit-field
         ASSUME(type_details->arena);
         ^
./perl.h:4227:36: note: expanded from macro 'ASSUME'
        ( (void) ({     __typeof__ (R) R__assume_ = (R);          \
                                   ^
1 error generated.
*** Error code 1

Stop.
make: stopped in /usr/home/jkeenan/gitwork/perl
```

jkeenan commented:
Okay, commit 0fa4efc074 in pr3 branch now builds.  Thanks.
